# IBM Data Science Specialization Capstone



## Segmenting and Clustering Neighborhoods in Toronto - Part 3

In Part 1 of this project, we collected the Toronto neighborhoods data from a web page, converted it to a pandas dataframe, and cleaned and pre-processed it.

In Part 2, we collected the latitude and logitude of each neighborhood and added it to our dataframe so that we can utilize the Foursquare location data.

In this final part, we will explore and cluster the neighborhoods in Toronto.


### Install required packages

In [1]:
import sys
!{sys.executable} -m pip install pandas
!{sys.executable} -m pip install beautifulsoup4
!{sys.executable} -m pip install requests
!{sys.executable} -m pip install geopy
!{sys.executable} -m pip install sklearn
!{sys.executable} -m pip install folium
!{sys.executable} -m pip install matplotlib

  Stored in directory: /home/dsxuser/.cache/pip/wheels/76/03/bb/589d421d27431bcd2c6da284d5f2286c8e3b2ea3cf1594c074
Successfully built sklearn
     |████████████████████████████████| 102kB 13.6MB/s ta 0:00:01



### Import dependencies

In [2]:
from bs4 import BeautifulSoup
from geopy.geocoders import Nominatim
from sklearn.cluster import KMeans
import requests
import pandas as pd
from pandas.io.json import json_normalize
import numpy as np
import folium
import matplotlib.cm as cm
import matplotlib.colors as colors


## Part 1 Revisited - Collect neighborhood data for Toronto and convert it into a pandas dataframe

The data will be collected from Wikipedia page: https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

This page contains a list of postal codes in Canada where the first letter is M.

In [3]:
# Get the HTML text from the wiki page
wiki_url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
html_text = requests.get(wiki_url).text

# Extract the table data using BeautifulSoup
soup = BeautifulSoup(html_text)
table = soup.find('table', attrs={'class':'wikitable sortable'})
trs = table.find_all('tr')

# Extract the text from all the table cells and add all rows
# to a list of rows.
rows = list()
for tr in trs:
    td = tr.find_all('td')
    row = [ele.text.strip() for ele in td]
    if row:
        # Ignore empty rows with no 'td',
        # applicable for the column headers row.
        rows.append(row)
        
# Convert the data to a pandas dataframe with 
# columns 'PostalCode', 'Borough', and 'Neighborhood'.
df = pd.DataFrame(rows,
                  columns=['PostalCode', 'Borough', 'Neighborhood'])

# Remove rows with column 'Not assigned' for column 'Borough'
df = df[df.Borough != 'Not assigned']
df.reset_index(inplace=True, drop=True)

# If a cell has a borough but a Not assigned neighborhood, 
# then the neighborhood will be the same as the borough.
# So, replace 'Neighborhood' columns with value as
# 'Not assigned' with the value of its 'Borough'
df['Neighborhood'] = df.apply(
    lambda row: 
    row['Borough'] if row['Neighborhood'] == 'Not assigned' 
    else row['Neighborhood'],
    axis=1)

# More than one neighborhood can exist in one postal code area.
# Combine rows with the same postal code into a single row.
# For example, in the table on the Wikipedia page, M5A is 
# listed twice and has two neighborhoods: Harbourfront and
# Regent Park. These two rows will be combined into one row
# with the neighborhoods separated with a comma.
df = df.groupby(['PostalCode', 'Borough'])['Neighborhood'].\
    apply(', '.join).to_frame()
df.reset_index(inplace=True)

In [4]:
df.head(12)

PostalCode      Borough                                       Neighborhood
0         M1B  Scarborough                                     Malvern, Rouge
1         M1C  Scarborough             Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough                  Guildwood, Morningside, West Hill
3         M1G  Scarborough                                             Woburn
4         M1H  Scarborough                                          Cedarbrae
5         M1J  Scarborough                                Scarborough Village
6         M1K  Scarborough        Kennedy Park, Ionview, East Birchmount Park
7         M1L  Scarborough                    Golden Mile, Clairlea, Oakridge
8         M1M  Scarborough    Cliffside, Cliffcrest, Scarborough Village West
9         M1N  Scarborough                        Birch Cliff, Cliffside West
10        M1P  Scarborough  Dorset Park, Wexford Heights, Scarborough Town...
11        M1R  Scarborough                                  Wexford, Maryvale

In [5]:
print("Dataframe shape: ", df.shape)

Dataframe shape:  (103, 3)



## Part 2 Revisited - Add latitude and longitude data to the dataframe

We will be using the CSV data at https://cocl.us/Geospatial_data to get the latitude and logitude information for all the postal codes.

Once we have the data, it will be concatenated with our original dataframe.

In [6]:
# Read the CSV data to a pandas dataframe.
geospatial_data = pd.read_csv('https://cocl.us/Geospatial_data')

## Combine the two dataframes.

# Concatenate the two dataframes using column "PostalCode" from 'df' and
# "Postal Code" from 'geospatial_data'.
df = pd.concat(
    [df.set_index('PostalCode'), geospatial_data.set_index('Postal Code')],
    axis=1, join='inner')

# Postal code will now be the index, change it to a non-index 
# column and reset index.
df.reset_index(inplace=True)

# Rename the 'index' column to 'PostalCode'
df.rename(columns={'index': 'PostalCode'}, inplace=True)


In [7]:
df.head(12)

PostalCode      Borough                                       Neighborhood  \
0         M1B  Scarborough                                     Malvern, Rouge   
1         M1C  Scarborough             Rouge Hill, Port Union, Highland Creek   
2         M1E  Scarborough                  Guildwood, Morningside, West Hill   
3         M1G  Scarborough                                             Woburn   
4         M1H  Scarborough                                          Cedarbrae   
5         M1J  Scarborough                                Scarborough Village   
6         M1K  Scarborough        Kennedy Park, Ionview, East Birchmount Park   
7         M1L  Scarborough                    Golden Mile, Clairlea, Oakridge   
8         M1M  Scarborough    Cliffside, Cliffcrest, Scarborough Village West   
9         M1N  Scarborough                        Birch Cliff, Cliffside West   
10        M1P  Scarborough  Dorset Park, Wexford Heights, Scarborough Town...   
11        M1R  Scarborough                                  Wexford, Maryvale   

     Latitude  Longitude  
0   43.806686 -79.194353  
1   43.784535 -79.160497  
2   43.763573 -79.188711  
3   43.770992 -79.216917  
4   43.773136 -79.239476  
5   43.744734 -79.239476  
6   43.727929 -79.262029  
7   43.711112 -79.284577  
8   43.716316 -79.239476  
9   43.692657 -79.264848  
10  43.757410 -79.273304  
11  43.750072 -79.295849

## Part 3 - Segment and Cluster Neighborhoods in Toronto

### 1. Explore the data

#### a. Get Toronto's latitude and longitude using geopy

In [8]:
address = 'Toronto, Canada'
geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
print('The geograpical coordinate of Toronto are {}, {}.'.format(
    location.latitude, location.longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### b. Create a map of Toronto using folium

In [9]:
map_toronto = folium.Map(
    location=[location.latitude, location.longitude],
    zoom_start=10)
map_toronto


#### c. Superimpose the neighborhoods on Toronto's map

Markers are added to the map for each of the neighborhoods and each marker is labeled in the format: "PostalCode (Neighborhood), Borough"

In [10]:
for _, row in df.iterrows():
    label = '{} ({}), {}'.format(
        row.PostalCode, row.Neighborhood, row.Borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [row.Latitude, row.Longitude],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto) 
    
map_toronto

#### d. Limit data set to boroughs with term 'Toronto'

For this lab, we will just pick neighborhoods in boroughs that contain the word "Toronto" in them.

In [11]:
# Create dataframe with boroughs containing the term 'Toronto'
toronto_data = df[df.Borough.str.contains('Toronto')]
toronto_data.reset_index(inplace=True, drop=True)
toronto_data

PostalCode           Borough  \
0         M4E      East Toronto   
1         M4K      East Toronto   
2         M4L      East Toronto   
3         M4M      East Toronto   
4         M4N   Central Toronto   
5         M4P   Central Toronto   
6         M4R   Central Toronto   
7         M4S   Central Toronto   
8         M4T   Central Toronto   
9         M4V   Central Toronto   
10        M4W  Downtown Toronto   
11        M4X  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
14        M5B  Downtown Toronto   
15        M5C  Downtown Toronto   
16        M5E  Downtown Toronto   
17        M5G  Downtown Toronto   
18        M5H  Downtown Toronto   
19        M5J  Downtown Toronto   
20        M5K  Downtown Toronto   
21        M5L  Downtown Toronto   
22        M5N   Central Toronto   
23        M5P   Central Toronto   
24        M5R   Central Toronto   
25        M5S  Downtown Toronto   
26        M5T  Downtown Toronto   
27        M5V  Downtown Toronto   
28        M5W  Downtown Toronto   
29        M5X  Downtown Toronto   
30        M6G  Downtown Toronto   
31        M6H      West Toronto   
32        M6J      West Toronto   
33        M6K      West Toronto   
34        M6P      West Toronto   
35        M6R      West Toronto   
36        M6S      West Toronto   
37        M7A  Downtown Toronto   
38        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                                         The Beaches  43.676357 -79.293031  
1                        The Danforth West, Riverdale  43.679557 -79.352188  
2                      India Bazaar, The Beaches West  43.668999 -79.315572  
3                                     Studio District  43.659526 -79.340923  
4                                       Lawrence Park  43.728020 -79.388790  
5                                    Davisville North  43.712751 -79.390197  
6                  North Toronto West,  Lawrence Park  43.715383 -79.405678  
7                                          Davisville  43.704324 -79.388790  
8                         Moore Park, Summerhill East  43.689574 -79.383160  
9   Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
10                                           Rosedale  43.679563 -79.377529  
11                        St. James Town, Cabbagetown  43.667967 -79.367675  
12                               Church and Wellesley  43.665860 -79.383160  
13                          Regent Park, Harbourfront  43.654260 -79.360636  
14                           Garden District, Ryerson  43.657162 -79.378937  
15                                     St. James Town  43.651494 -79.375418  
16                                        Berczy Park  43.644771 -79.373306  
17                                 Central Bay Street  43.657952 -79.387383  
18                           Richmond, Adelaide, King  43.650571 -79.384568  
19  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
20           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
21                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
22                                           Roselawn  43.711695 -79.416936  
23    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                     University of Toronto, Harbord  43.662696 -79.400049  
26          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
27  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
28                                     Stn A PO Boxes  43.646435 -79.374846  
29             First Canadian Place, Underground city  43.648429 -79.382280  
30                                           Christie  43.669542 -79.422564  
31                       Dufferin, Dovercourt Village  43.669005 -79.442259  
32                          

In [12]:
toronto_data.shape

(39, 5)

So, there are 39 neighborhoods with boroughs containing the term 'Toronto'.

Let us visualize these on Toronto's map.

In [13]:

# Create a map instance
map_toronto = folium.Map(
    location=[location.latitude, location.longitude],
    zoom_start=11)

# Add markers for all 38 neighborhoods
for _, row in toronto_data.iterrows():
    label = '{} ({}), {}'.format(
        row.PostalCode, row.Neighborhood, row.Borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [row.Latitude, row.Longitude],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto) 
    
map_toronto

#### e. Define Foursquare crendentials

To avoid exposing the credentials, I have saved the client ID and secret in file "config.py" and placed it alongside this notebook.

In [23]:

CLIENT_ID ='W3KIP5CE51Q3RMG5JLCWZTHF3VOEA3TRMKB4ANGBIGOAFIFF'
CLIENT_SECRET = 'QMWCEB0JGHD4FIX1MIMPBGZ5I0GDTQL53KY4PUGXZPBMRDBA'
VERSION = '20180605'


#### f. Explore a neighborhood in Toronto

Now let us explore the first neighborhood in our dataframe.

Get the first neighborhood's name, latitude and longitude.

In [24]:

neighborhood_name = toronto_data.loc[0, 'Neighborhood']
neighborhood_latitude = toronto_data.loc[0, 'Latitude']
neighborhood_longitude = toronto_data.loc[0, 'Longitude']

print("Latitude and longitude of neighborhood '{}' are [{}, {}]".format(
    neighborhood_name, neighborhood_latitude, neighborhood_longitude))

Latitude and longitude of neighborhood 'The Beaches' are [43.67635739999999, -79.2930312]


#### Get the top 100 venues in 'The Beaches' within a radius of 500 meters.

In [25]:
# Construct the URL
limit = 100
radius = 500
explore_url_prefix = 'https://api.foursquare.com/v2/venues/explore'
url = '{}?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    explore_url_prefix, CLIENT_ID, CLIENT_SECRET, VERSION, 
    neighborhood_latitude, neighborhood_longitude, radius, limit)

In [26]:
# Get the venues.
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ed35eed9c6f59001b464c71'},
 'response': {'headerLocation': 'The Beaches',
  'headerFullLocation': 'The Beaches, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 6,
  'suggestedBounds': {'ne': {'lat': 43.680857404499996,
    'lng': -79.28682091449052},
   'sw': {'lat': 43.67185739549999, 'lng': -79.29924148550948}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bd461bc77b29c74a07d9282',
       'name': 'Glen Manor Ravine',
       'location': {'address': 'Glen Manor',
        'crossStreet': 'Queen St.',
        'lat': 43.67682094413784,
        'lng': -79.29394208780985,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.67682094413784,
          'lng': -79.29394208780985}],
        'distanc

In [27]:
# Explore the venues
venues = results['response']['groups'][0]['items']
venues

[{'reasons': {'count': 0,
   'items': [{'summary': 'This spot is popular',
     'type': 'general',
     'reasonName': 'globalInteractionReason'}]},
  'venue': {'id': '4bd461bc77b29c74a07d9282',
   'name': 'Glen Manor Ravine',
   'location': {'address': 'Glen Manor',
    'crossStreet': 'Queen St.',
    'lat': 43.67682094413784,
    'lng': -79.29394208780985,
    'labeledLatLngs': [{'label': 'display',
      'lat': 43.67682094413784,
      'lng': -79.29394208780985}],
    'distance': 89,
    'cc': 'CA',
    'city': 'Toronto',
    'state': 'ON',
    'country': 'Canada',
    'formattedAddress': ['Glen Manor (Queen St.)', 'Toronto ON', 'Canada']},
   'categories': [{'id': '4bf58dd8d48988d159941735',
     'name': 'Trail',
     'pluralName': 'Trails',
     'shortName': 'Trail',
     'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/parks_outdoors/hikingtrail_',
      'suffix': '.png'},
     'primary': True}],
   'photos': {'count': 0, 'groups': []}},
  'referralId': 'e-0-4bd461bc77b2

In [28]:
# Normalize the JSON response
neighborhood_venues = json_normalize(venues)
neighborhood_venues

reasons.count                                      reasons.items  \
0              0  [{'summary': 'This spot is popular', 'type': '...   
1              0  [{'summary': 'This spot is popular', 'type': '...   
2              0  [{'summary': 'This spot is popular', 'type': '...   
3              0  [{'summary': 'This spot is popular', 'type': '...   
4              0  [{'summary': 'This spot is popular', 'type': '...   
5              0  [{'summary': 'This spot is popular', 'type': '...   

                       referralId  \
0  e-0-4bd461bc77b29c74a07d9282-0   
1  e-0-4ad4c062f964a52011f820e3-1   
2  e-0-4b8daea1f964a520480833e3-2   
3  e-0-4dbc8fe96a23e294ba3237bd-3   
4  e-0-4df91c4bae60f95f82229ad5-4   
5  e-0-4bc7fcce6501c9b6bf813f29-5   

                                    venue.categories  \
0  [{'id': '4bf58dd8d48988d159941735', 'name': 'T...   
1  [{'id': '50aa9e744b90af0d42d5de0e', 'name': 'H...   
2  [{'id': '4bf58dd8d48988d11b941735', 'name': 'P...   
3  [{'id': '4bf58dd8d48988d163941735', 'name': 'P...   
4  [{'id': '4f2a25ac4b909258e854f55f', 'name': 'N...   
5  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...   

                   venue.id venue.location.address venue.location.cc  \
0  4bd461bc77b29c74a07d9282             Glen Manor                CA   
1  4ad4c062f964a52011f820e3       125 Southwood Dr                CA   
2  4b8daea1f964a520480833e3       676 Kingston Rd.                CA   
3  4dbc8fe96a23e294ba3237bd   131 Glen Manor Drive                CA   
4  4df91c4bae60f95f82229ad5                    NaN                CA   
5  4bc7fcce6501c9b6bf813f29      663 Kingston Road                CA   

  venue.location.city venue.location.country venue.location.crossStreet  ...  \
0             Toronto                 Canada                  Queen St.  ...   
1             Toronto                 Canada                        NaN  ...   
2             Toronto                 Canada                at Main St.  ...   
3             Toronto                 Canada                        NaN  ...   
4             Toronto                 Canada                        NaN  ...   
5             Toronto                 Canada                    Main St  ...   

                     venue.location.formattedAddress  \
0       [Glen Manor (Queen St.), Toronto ON, Canada]   
1     [125 Southwood Dr, Toronto ON M4E 0B8, Canada]   
2  [676 Kingston Rd. (at Main St.), Toronto ON M4...   
3         [131 Glen Manor Drive, Toronto ON, Canada]   
4                               [Toronto ON, Canada]   
5  [663 Kingston Road (Main St), Toronto ON, Canada]   

                       venue.location.labeledLatLngs venue.location.lat  \
0  [{'label': 'display', 'lat': 43.67682094413784...          43.676821   
1  [{'label': 'display', 'lat': 43.678879, 'lng':...          43.678879   
2  [{'label': 'display', 'lat': 43.67918143494101...          43.679181   
3  [{'label': 'display', 'lat': 43.67527822698259...          43.675278   
4  [{'label': 'display', 'lat': 43.68056321147582...          43.680563   
5  [{'label': 'display', 'lat': 43.67889707815811...          43.678897   

   venue.location.lng  venue.location.postalCode venue.location.state  \
0          -79.293942                        NaN                   ON   
1          -79.297734                    M4E 0B8                   ON   
2          -79.297215                    M4E 1R4                   ON   
3          -79.294647                        NaN                   ON   
4          -79.292869                        NaN                   ON   
5          -79.297745                        NaN                   ON   

                           venue.name venue.photos.count  venue.photos.groups  \
0                   Glen Manor Ravine                  0                   []   
1  The Big Carrot Natural Food Market                  0                   []   
2                 Grover Pub and Grub                  0                   []   
3                   Glen Stewar

This neighborhood only seems to have 6 venues. Let us go ahead and explore it further.

In [29]:

# Filter out the venue name, category, latitude and logitude.
venue_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
neighborhood_venues = neighborhood_venues.loc[:, venue_columns]
neighborhood_venues

venue.name  \
0                   Glen Manor Ravine   
1  The Big Carrot Natural Food Market   
2                 Grover Pub and Grub   
3                   Glen Stewart Park   
4                       Upper Beaches   
5                          Dip 'n Sip   

                                    venue.categories  venue.location.lat  \
0  [{'id': '4bf58dd8d48988d159941735', 'name': 'T...           43.676821   
1  [{'id': '50aa9e744b90af0d42d5de0e', 'name': 'H...           43.678879   
2  [{'id': '4bf58dd8d48988d11b941735', 'name': 'P...           43.679181   
3  [{'id': '4bf58dd8d48988d163941735', 'name': 'P...           43.675278   
4  [{'id': '4f2a25ac4b909258e854f55f', 'name': 'N...           43.680563   
5  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...           43.678897   

   venue.location.lng  
0          -79.293942  
1          -79.297734  
2          -79.297215  
3          -79.294647  
4          -79.292869  
5          -79.297745

In [30]:

# Change the column names to just the last part after the '.'.
neighborhood_venues.columns = [column.split(".")[-1] for column in neighborhood_venues.columns]
neighborhood_venues

name  \
0                   Glen Manor Ravine   
1  The Big Carrot Natural Food Market   
2                 Grover Pub and Grub   
3                   Glen Stewart Park   
4                       Upper Beaches   
5                          Dip 'n Sip   

                                          categories        lat        lng  
0  [{'id': '4bf58dd8d48988d159941735', 'name': 'T...  43.676821 -79.293942  
1  [{'id': '50aa9e744b90af0d42d5de0e', 'name': 'H...  43.678879 -79.297734  
2  [{'id': '4bf58dd8d48988d11b941735', 'name': 'P...  43.679181 -79.297215  
3  [{'id': '4bf58dd8d48988d163941735', 'name': 'P...  43.675278 -79.294647  
4  [{'id': '4f2a25ac4b909258e854f55f', 'name': 'N...  43.680563 -79.292869  
5  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...  43.678897 -79.297745

In [31]:

# Extract the category names from a row.
# We will get the first item in the categories list and then get its name.

def get_category(row):
    categories_list = row['categories']
    if categories_list:
        return categories_list[0]['name']
    return None

In [32]:
# Replace the values in categories column with the first catogory name.
neighborhood_venues['categories'] = neighborhood_venues.apply(get_category, axis=1)
neighborhood_venues

name         categories        lat        lng
0                   Glen Manor Ravine              Trail  43.676821 -79.293942
1  The Big Carrot Natural Food Market  Health Food Store  43.678879 -79.297734
2                 Grover Pub and Grub                Pub  43.679181 -79.297215
3                   Glen Stewart Park               Park  43.675278 -79.294647
4                       Upper Beaches       Neighborhood  43.680563 -79.292869
5                          Dip 'n Sip        Coffee Shop  43.678897 -79.297745

#### g. Explore neighborhoods in Toronto

Now we will explore all the neighborhoods in our dataframe. We will repeat the process that we did for 'The Beaches' neighborhood in the previous section.

In [33]:
venues_list = list()

for name, lat, lng in zip(toronto_data['Neighborhood'], toronto_data['Latitude'], toronto_data['Longitude']):
    print("Collecting venues for neighborhood:", name)
    
    # Create API request URL
    limit = 100
    radius = 500
    explore_url_prefix = 'https://api.foursquare.com/v2/venues/explore'
    url = '{}?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
        explore_url_prefix, CLIENT_ID, CLIENT_SECRET, VERSION, 
        lat, lng, radius, limit)
    
    # Make the request
    neighborhood_venues = requests.get(url).json()["response"]['groups'][0]['items']
    
    # Add relevant info to venues_list
    venues_list.extend([(
        name, lat, lng,
        v['venue']['name'], 
        v['venue']['location']['lat'], 
        v['venue']['location']['lng'],  
        v['venue']['categories'][0]['name']) for v in neighborhood_venues])

print("Done")


Done


In [34]:

# Let us see how many venues we have got.
len(venues_list)

1616

In [35]:

# Convert the venues_list to a dataframe.
toronto_venues = pd.DataFrame(venues_list)
toronto_venues.columns = [
    'Neighborhood', 'Neighborhood Latitude', 'Neighborhood Longitude', 
    'Venue', 'Venue Latitude', 'Venue Longitude', 'Venue Category']

In [36]:
print(toronto_venues.shape)
toronto_venues.head()

(1616, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                   Glen Stewart Park       43.675278       -79.294647   
4                       Upper Beaches       43.680563       -79.292869   

      Venue Category  
0              Trail  
1  Health Food Store  
2                Pub  
3               Park  
4       Neighborhood

Let us check the number of venues in each neighborhood.

In [37]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            58   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     17   
CN Tower, King and Spadina, Railway Lands, Harb...                     14   
Central Bay Street                                                     62   
Christie                                                               17   
Church and Wellesley                                                   77   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                        7   
Dufferin, Dovercourt Village                                           16   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          24   
India Bazaar, The Beaches West                                         22   
Kensington Market, Chinatown, Grange Park                              58   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               44   
Moore Park, Summerhill East                                             3   
North Toronto West,  Lawrence Park                                     22   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            36   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                               91   
Rosedale                                                                4   
Roselawn                                                                3   
Runnymede, Swansea                                                     42   
St. James Town                                                         80   
St. James Town, Cabbagetown                                            42   
Stn A PO Boxes                                                         93   
Studio District                                                        41   
Summerhill West, Rathnelly, South Hill, Forest ...                     17   
The Annex, North Midtown, Yorkville                                    22   
The Beaches                                                             6   
The Danforth West, Riverdale                                           42   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             58   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      17   
CN Tower, King and Spadina, Railway Lands, Harb...                      14   
Central Bay Street                                                      62   
Christie                                                                17   
Church and Wellesley                                                    77   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              33   
Davisville North                    

Let us find how many unique venue categories we have.

In [38]:
print('There are {} uniques categories'.format(
    len(toronto_venues['Venue Category'].unique())))

There are 236 uniques categories


### 2. Analyze each neighborhood

In [39]:
# One hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot.head()

Afghan Restaurant  Airport  Airport Food Court  Airport Lounge  \
0                  0        0                   0               0   
1                  0        0                   0               0   
2                  0        0                   0               0   
3                  0        0                   0               0   
4                  0        0                   0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  Art Gallery  ...  Toy / Game Store  Trail  Train Station  \
0         0            0  ...                 0      1              0   
1         0            0  ...                 0      0              0   
2         0            0  ...                 0      0              0   
3         0            0  ...                 0      0              0   
4         0            0  ...                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0         0            0              0            0  
1         0            0              0            0  
2         0            0              0            0  
3         0            0              0            0  
4         0            0              0            0  

[5 rows x 236 columns]

In [40]:
# Add neighborhood column back to dataframe as NeighborhoodName
toronto_onehot.insert(0, 'NeighborhoodName', toronto_venues['Neighborhood']) 
toronto_onehot.head()

NeighborhoodName  Afghan Restaurant  Airport  Airport Food Court  \
0      The Beaches                  0        0                   0   
1      The Beaches                  0        0                   0   
2      The Beaches                  0        0                   0   
3      The Beaches                  0        0                   0   
4      The Beaches                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Toy / Game Store  Trail  Train Station  \
0             0         0  ...                 0      1              0   
1             0         0  ...                 0      0              0   
2             0         0  ...                 0      0              0   
3             0         0  ...                 0      0              0   
4             0         0  ...                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0         0            0              0            0  
1         0            0              0            0  
2         0            0              0            0  
3         0            0              0            0  
4         0            0              0            0  

[5 rows x 237 columns]

In [41]:
toronto_onehot.shape

(1616, 237)

In [42]:
# Check the size of this dataframe.
venues_count, categories_count = toronto_onehot.shape
print("So we have {} different venues and {} venue categories.".format(
    venues_count, categories_count))

So we have 1616 different venues and 237 venue categories.


### Now we will group the rows by neighborhood name and by calculating the mean of the frequency of occurence of each category.

In [44]:
toronto_grouped = toronto_onehot.groupby('NeighborhoodName').mean().reset_index()
toronto_grouped

NeighborhoodName  Afghan Restaurant  \
0                                         Berczy Park           0.000000   
1        Brockton, Parkdale Village, Exhibition Place           0.000000   
2   Business reply mail Processing Centre, South C...           0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
4                                  Central Bay Street           0.000000   
5                                            Christie           0.000000   
6                                Church and Wellesley           0.012987   
7                      Commerce Court, Victoria Hotel           0.000000   
8                                          Davisville           0.000000   
9                                    Davisville North           0.000000   
10                       Dufferin, Dovercourt Village           0.000000   
11             First Canadian Place, Underground city           0.000000   
12    Forest Hill North & West, Forest Hill Road Park           0.000000   
13                           Garden District, Ryerson           0.000000   
14  Harbourfront East, Union Station, Toronto Islands           0.000000   
15                      High Park, The Junction South           0.000000   
16                     India Bazaar, The Beaches West           0.000000   
17          Kensington Market, Chinatown, Grange Park           0.000000   
18                                      Lawrence Park           0.000000   
19                           Little Portugal, Trinity           0.000000   
20                        Moore Park, Summerhill East           0.000000   
21                 North Toronto West,  Lawrence Park           0.000000   
22                             Parkdale, Roncesvalles           0.000000   
23        Queen's Park, Ontario Provincial Government           0.000000   
24                          Regent Park, Harbourfront           0.000000   
25                           Richmond, Adelaide, King           0.000000   
26                                           Rosedale           0.000000   
27                                           Roselawn           0.000000   
28                                 Runnymede, Swansea           0.000000   
29                                     St. James Town           0.000000   
30                        St. James Town, Cabbagetown           0.000000   
31                                     Stn A PO Boxes           0.000000   
32                                    Studio District           0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...           0.000000   
34                The Annex, North Midtown, Yorkville           0.000000   
35                                        The Beaches           0.000000   
36                       The Danforth West, Riverdale           0.000000   
37           Toronto Dominion Centre, Design Exchange           0.000000   
38                     University of Toronto, Harbord           0.000000   

     Airport  Airport Food Court  Airport Lounge  Airport Service  \
0   0.000000            0.000000        0.000000         0.000000   
1   0.000000            0.000000        0.000000         0.000000   
2   0.000000            0.000000        0.000000         0.000000   
3   0.071429            0.071429        0.142857         0.142857   
4   0.000000            0.000000        0.000000         0.000000   
5   0.000000            0.000000        0.000000         0.000000   
6   0.000000            0.000000        0.000000         0.000000   
7   0.000000            0.000000        0.000000         0.000000   
8   0.000000            0.000000        0.000000         0.000000   
9   0.000000            0.000000        0.000000         0.000000   
10  0.000000            0.000000        0.000000         0.000000   
11  0.000000            0.000000        0.000000         0.000000   
12  0.000000            0.000000        0.000000         0.000000   
13  0.000000            0.0000


The dataframe has 39 rows and 237 columns, which is as expected,i.e., 39 neighborhoods and 237 categories.

Now we will print each of the neighborhoods with the top 5 most common venues in them

In [45]:
num_top_venues = 5
for neighborhood in toronto_grouped['NeighborhoodName']:
    print("------{}------".format(neighborhood))
    temp = toronto_grouped[toronto_grouped['NeighborhoodName'] == neighborhood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

------Berczy Park------
          venue  freq
0   Coffee Shop  0.07
1  Cocktail Bar  0.05
2   Cheese Shop  0.03
3           Pub  0.03
4        Bakery  0.03


------Brockton, Parkdale Village, Exhibition Place------
            venue  freq
0            Café  0.13
1  Breakfast Spot  0.09
2       Nightclub  0.09
3     Coffee Shop  0.09
4          Bakery  0.04


------Business reply mail Processing Centre, South Central Letter Processing Plant Toronto------
                  venue  freq
0    Light Rail Station  0.12
1           Yoga Studio  0.06
2                   Spa  0.06
3         Garden Center  0.06
4  Fast Food Restaurant  0.06


------CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport------
              venue  freq
0    Airport Lounge  0.14
1   Airport Service  0.14
2  Airport Terminal  0.14
3   Harbor / Marina  0.07
4           Airport  0.07


------Central Bay Street------
                 venue  freq
0          Coffee Shop 

### Next we will create a pandas dataframe with the top 10 most common venues in each neighborhood

Function to sort venues in descending order of their frequency.

In [46]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]



Now we will create the dataframe.

In [47]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['NeighborhoodName']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(
        toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood 1st Most Common Venue  \
0                                         Berczy Park           Coffee Shop   
1        Brockton, Parkdale Village, Exhibition Place                  Café   
2   Business reply mail Processing Centre, South C...    Light Rail Station   
3   CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                  Central Bay Street           Coffee Shop   
5                                            Christie         Grocery Store   
6                                Church and Wellesley           Coffee Shop   
7                      Commerce Court, Victoria Hotel           Coffee Shop   
8                                          Davisville          Dessert Shop   
9                                    Davisville North                   Gym   
10                       Dufferin, Dovercourt Village                Bakery   
11             First Canadian Place, Underground city           Coffee Shop   
12    Forest Hill North & West, Forest Hill Road Park                 Trail   
13                           Garden District, Ryerson        Clothing Store   
14  Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
15                      High Park, The Junction South   Arts & Crafts Store   
16                     India Bazaar, The Beaches West        Sandwich Place   
17          Kensington Market, Chinatown, Grange Park                  Café   
18                                      Lawrence Park                  Park   
19                           Little Portugal, Trinity                   Bar   
20                        Moore Park, Summerhill East            Playground   
21                 North Toronto West,  Lawrence Park        Clothing Store   
22                             Parkdale, Roncesvalles        Breakfast Spot   
23        Queen's Park, Ontario Provincial Government           Coffee Shop   
24                          Regent Park, Harbourfront           Coffee Shop   
25                           Richmond, Adelaide, King           Coffee Shop   
26                                           Rosedale                  Park   
27                                           Roselawn        Ice Cream Shop   
28                                 Runnymede, Swansea           Coffee Shop   
29                                     St. James Town           Coffee Shop   
30                        St. James Town, Cabbagetown           Coffee Shop   
31                                     Stn A PO Boxes           Coffee Shop   
32                                    Studio District                  Café   
33  Summerhill West, Rathnelly, South Hill, Forest...           Coffee Shop   
34                The Annex, North Midtown, Yorkville        Sandwich Place   
35                                        The Beaches                  Park   
36                       The Danforth West, Riverdale      Greek Restaurant   
37           Toronto Dominion Centre, Design Exchange           Coffee Shop   
38                     University of Toronto, Harbord                  Café   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Cocktail Bar    Seafood Restaurant                  Café   
1         Breakfast Spot             Nightclub           Coffee Shop   
2          Auto Workshop                  Park            Comic Shop   
3        Airport Service      Airport Terminal         Boat or Ferry   
4     Italian Restaurant        Sandwich Place                  Café   
5                   Café                  Park           Candy Store   
6    Japanese Restaurant      Sushi Restaurant            Restaurant   
7             Restaurant                  Café                 Hotel   
8         Sandwich Place                  Café      Sushi Restaurant   
9         Breakfast Spot        Sandwich Place                 Hotel   
10              Pharmacy            Smoke Shop           Music Venue   
11              

## 3. Cluster the Toronto Neighborhoods

We will run k-means and cluster the neighborhoods into 5 clusters.

In [48]:
# Set number of clusters
k = 5

# Drop the neighborhood name column so that each column contains only the feature set.
toronto_grouped_clustering = toronto_grouped.drop('NeighborhoodName', 1)

# Run k-means clustering
kmeans = KMeans(n_clusters=k, random_state=0).fit(toronto_grouped_clustering)

# Check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

##### Now we will create a dataframe with the neighborhood information, top 10 common venues as well as the cluster labels.

In [49]:
# Add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# Merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

print(toronto_merged.shape)
toronto_merged.head()

(39, 16)


PostalCode          Borough                    Neighborhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               0                  Park                   Pub   
1 -79.352188               0      Greek Restaurant    Italian Restaurant   
2 -79.315572               0        Sandwich Place  Fast Food Restaurant   
3 -79.340923               0                  Café           Coffee Shop   
4 -79.388790               1                  Park              Bus Line   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Coffee Shop     Health Food Store          Neighborhood   
1           Coffee Shop            Restaurant        Ice Cream Shop   
2           Pizza Place      Sushi Restaurant                   Pub   
3             Gastropub   American Restaurant                Bakery   
4           Swim School           Yoga Studio          Dessert Shop   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                   Trail      Department Store          Dessert Shop   
1  Furniture / Home Store           Yoga Studio          Liquor Store   
2           Burrito Place            Board Shop    Italian Restaurant   
3                 Brewery           Yoga Studio           Pizza Place   
4             Event Space  Ethiopian Restaurant     Electronics Store   

         9th Most Common Venue 10th Most Common Venue  
0                        Diner         Discount Store  
1                          Spa              Juice Bar  
2            Fish & Chips Shop             Steakhouse  
3                    Pet Store                   Park  
4  Eastern European Restaurant             Donut Shop

### Visualize the clusters

In [50]:
# Create a map instance
map_toronto = folium.Map(
    location=[location.latitude, location.longitude],
    zoom_start=11)

# Set color scheme for the clusters
x = np.arange(k)
ys = [i + x + (i*x)**2 for i in range(k)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# Add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(
        toronto_merged['Latitude'], toronto_merged['Longitude'],
        toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_toronto)
       
map_toronto

### 4. Examine the clusters

Now let us examine the clusters and see how they differ from each other in terms of popular venues.

#### Cluster 0

In [51]:
cluster_0 = toronto_merged.loc[toronto_merged['Cluster Labels'] == 0,
                               toronto_merged.columns[
                                   [2] + list(range(
                                       5, toronto_merged.shape[1]))]]
cluster_0

Neighborhood  Cluster Labels  \
0                                         The Beaches               0   
1                        The Danforth West, Riverdale               0   
2                      India Bazaar, The Beaches West               0   
3                                     Studio District               0   
5                                    Davisville North               0   
6                  North Toronto West,  Lawrence Park               0   
7                                          Davisville               0   
9   Summerhill West, Rathnelly, South Hill, Forest...               0   
11                        St. James Town, Cabbagetown               0   
12                               Church and Wellesley               0   
13                          Regent Park, Harbourfront               0   
14                           Garden District, Ryerson               0   
15                                     St. James Town               0   
16                                        Berczy Park               0   
17                                 Central Bay Street               0   
18                           Richmond, Adelaide, King               0   
19  Harbourfront East, Union Station, Toronto Islands               0   
20           Toronto Dominion Centre, Design Exchange               0   
21                     Commerce Court, Victoria Hotel               0   
24                The Annex, North Midtown, Yorkville               0   
25                     University of Toronto, Harbord               0   
26          Kensington Market, Chinatown, Grange Park               0   
27  CN Tower, King and Spadina, Railway Lands, Har...               0   
28                                     Stn A PO Boxes               0   
29             First Canadian Place, Underground city               0   
30                                           Christie               0   
31                       Dufferin, Dovercourt Village               0   
32                           Little Portugal, Trinity               0   
33       Brockton, Parkdale Village, Exhibition Place               0   
34                      High Park, The Junction South               0   
35                             Parkdale, Roncesvalles               0   
36                                 Runnymede, Swansea               0   
37        Queen's Park, Ontario Provincial Government               0   
38  Business reply mail Processing Centre, South C...               0   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                   Park                   Pub           Coffee Shop   
1       Greek Restaurant    Italian Restaurant           Coffee Shop   
2         Sandwich Place  Fast Food Restaurant           Pizza Place   
3                   Café           Coffee Shop             Gastropub   
5                    Gym        Breakfast Spot        Sandwich Place   
6         Clothing Store           Coffee Shop           Yoga Studio   
7           Dessert Shop        Sandwich Place                  Café   
9            Coffee Shop                  Bank                   Pub   
11           Coffee Shop                  Café            Restaurant   
12           Coffee Shop   Japanese Restaurant      Sushi Restaurant   
13           Coffee Shop                   Pub                Bakery   
14        Clothing Store           Coffee Shop   Japanese Restaurant   
15           Coffee Shop                  Café          Cocktail Bar   
16           Coffee Shop          Cocktail Bar    Seafood Restaurant   
17           Coffee Shop    Italian Restaurant        Sandwich Place   
18           Coffee Shop                  Café            Restaurant   
19           Coffee Shop              Aquarium                 Hotel   
20           Coffee Shop                  Café                 Hotel   
21           Coffee Shop            Restaurant                  Café   
24        Sandwich Place                  Caf

In [52]:
print(cluster_0.shape)

(34, 12)


#### Cluster 1

In [53]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1,
                   toronto_merged.columns[
                       [2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
4  Lawrence Park               1                  Park              Bus Line   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4           Swim School           Yoga Studio          Dessert Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
4           Event Space  Ethiopian Restaurant     Electronics Store   

         9th Most Common Venue 10th Most Common Venue  
4  Eastern European Restaurant             Donut Shop

#### Cluster 2

In [54]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2,
                   toronto_merged.columns[
                       [2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
10                                         Rosedale               2   
23  Forest Hill North & West, Forest Hill Road Park               2   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
10                  Park            Playground                 Trail   
23                 Trail                  Park      Sushi Restaurant   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
10           Yoga Studio         Deli / Bodega  Ethiopian Restaurant   
23         Jewelry Store      Doner Restaurant                 Diner   

   7th Most Common Venue        8th Most Common Venue 9th Most Common Venue  \
10     Electronics Store  Eastern European Restaurant            Donut Shop   
23        Discount Store          Distribution Center               Dog Run   

   10th Most Common Venue  
10       Doner Restaurant  
23            Yoga Studio

#### Cluster 3

In [55]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3,
                   toronto_merged.columns[
                       [2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
22     Roselawn               3        Ice Cream Shop          Home Service   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
22                Garden      Department Store    Falafel Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
22           Event Space  Ethiopian Restaurant     Electronics Store   

          9th Most Common Venue 10th Most Common Venue  
22  Eastern European Restaurant             Donut Shop

#### Cluster 4

In [56]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4,
                   toronto_merged.columns[
                       [2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
8  Moore Park, Summerhill East               4            Playground   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
8           Summer Camp            Restaurant           Yoga Studio   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
8         Deli / Bodega  Ethiopian Restaurant     Electronics Store   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
8  Eastern European Restaurant            Donut Shop       Doner Restaurant


## Insights:

There are 32 neighborhoods in the first cluster! It appears that coffee shops and cafes are very popular in the neighborhoods in this cluster and in general in Toronto!

The rest of the clusters are much smaller and coffee shops are not that common in those.

Out of the 5 clusters, 3 have only 1 neighborhood in them. These are the most distinct neighborhoods of all.


### Thank you!